File: algorithms/recursive_list_reverse/reverse.rb
#!/usr/bin/env ruby
# 2017 Daniel Varga (vargad88@gmail.com)

class LinkedList
    attr_reader :data
    attr_reader :next

    def initialize(data, rest=nil)
        @data = data
        @next = rest
    end

    def to_s
        return "%d, %s" % [self.data, self.next]
    end
end

def make_list(*elements)
    # alternate solution
    #return elements.reverse.inject(nil) { |c, e| LinkedList.new(e, c) }
    list = nil
    elements.reverse.each { |e| list = LinkedList(e, list) }
    return list
end

def reverse(list, rest=nil)
    return rest if list.nil?
    return reverse(list.next, LinkedList.new(list.data, rest))
end

l = make_list(1,2,3,4,5,6,7)
puts(l)
l2 = reverse(l)
puts(l2)

File: algorithms/recursive_list_reverse/reverse.cpp
// 2017 Daniel Varga (vargad88@gmail.com)

#include <iostream>
#include <memory>
#include <iterator>

struct LinkedList
{
    using NodePtr = std::unique_ptr<LinkedList>;

    LinkedList(int data, NodePtr &&next=nullptr)
        : data(data)
        , next(std::move(next))
    {}

    int data;
    NodePtr next;
};

std::ostream& operator<<(std::ostream &os, const std::unique_ptr<LinkedList> &list)
{
    os << list->data;
    if (list->next)
        os << ", " << list->next;
    return os;
}

LinkedList::NodePtr make_list(const std::initializer_list<int> &init_list)
{
    LinkedList::NodePtr list = nullptr;
    for (auto it = std::rbegin(init_list); it != std::rend(init_list); ++it)
        list = std::make_unique<LinkedList>(*it, std::move(list));
    return std::move(list);
}

LinkedList::NodePtr reverse(const LinkedList::NodePtr &list, LinkedList::NodePtr &&rest = nullptr)
{
    if (list == nullptr)
        return std::move(rest);
    return reverse(list->next, std::make_unique<LinkedList>(list->data, std::move(rest)));
}

int main()
{
    auto list = make_list({ 42, 43, 44, 45, 46, 47 });
    std::cout << list << std::endl;
    auto list2 = reverse(std::move(list));
    std::cout << list2 << std::endl;
    return 0;
}
